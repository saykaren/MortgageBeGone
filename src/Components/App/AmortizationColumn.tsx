import React from 'react';

interface AmortizationColumnProps {
  title: string;
  initialRow: string | number;
  arr: number[];
}

const AmortizationColumn = ({
  title,
  initialRow,
  arr,
}: AmortizationColumnProps) => {
  return (
    <div className="tableCell">
      {title}
      <div className="cellDetails">{initialRow}</div>
      {arr.map((value, index) => (
        <div className="cellDetails" key={index}>
          {value}
        </div>
      ))}
    </div>
  );
};

export default AmortizationColumn;
