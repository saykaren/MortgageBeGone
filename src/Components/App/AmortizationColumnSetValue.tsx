import React from "react";

interface AmortizationColumnSetValueProps {
    title: string;
    copyValue: number
    arr: number[];
}

const AmortizationColumnSetValue = ({title, copyValue, arr}: AmortizationColumnSetValueProps) => {
    return (
        <div className="tableCell">
            {title}
            {arr.map((value, index) => (
                <div className="cellDetails" key={index}>
                    {copyValue}
                </div>
            ))}
        </div>
    );
};

export default AmortizationColumnSetValue;