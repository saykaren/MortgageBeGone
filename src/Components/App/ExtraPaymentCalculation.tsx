
import NumberConverter from './NumberConverter';

interface ExtraPaymentCalculationProps {
    extraNewEndingPrincipalArray: number[];
    interestRate: number;
    setModal: (arg: boolean)=>void;
    extraPayment: number;
    monthlyPayment: number;
    extraPrincipalPaidArray: number[];
    setExtraPrincipalPaidArray: ([])=>void;
    extraInterestPaidArray: number[];
    setExtraInterestPaidArray: ([])=>void;
    setExtraNewEndingPrincipalArray: ([])=>void;
    principal: number;
}

const ExtraPaymentCalculation = ({
                                     extraNewEndingPrincipalArray,
                                     interestRate,
                                     setModal,
                                     extraPayment,
                                     monthlyPayment,
                                     extraPrincipalPaidArray,
                                     setExtraPrincipalPaidArray,
                                     extraInterestPaidArray,
                                     setExtraInterestPaidArray,
                                     setExtraNewEndingPrincipalArray,
                                     principal,
                                 } : ExtraPaymentCalculationProps) => {
    let currentExtraPrincipal: any;
    if (extraNewEndingPrincipalArray.length > 0) {
        currentExtraPrincipal = NumberConverter(
            extraNewEndingPrincipalArray[extraNewEndingPrincipalArray.length - 1],
        );
    }

    const processEachExtraPayment = (num: number) => {
        let paymentExtraInterestPaid = NumberConverter(
            currentExtraPrincipal * ((interestRate * 0.01) / 12),
        );
        let principalExtraPaid;
        ///Last payment 0
        if (num === 0) {
            principalExtraPaid = NumberConverter(currentExtraPrincipal);
            setModal(true);
        } else {
            principalExtraPaid = NumberConverter(
                extraPayment + monthlyPayment - paymentExtraInterestPaid,
            );
        }
        let extraBalance = NumberConverter(
            currentExtraPrincipal - principalExtraPaid,
        );
        setExtraPrincipalPaidArray([
            ...extraPrincipalPaidArray,
            principalExtraPaid,
        ]);
        setExtraInterestPaidArray([
            ...extraInterestPaidArray,
            paymentExtraInterestPaid,
        ]);
        setExtraNewEndingPrincipalArray([
            ...extraNewEndingPrincipalArray,
            extraBalance,
        ]);
    };

    switch (true) {
        case extraNewEndingPrincipalArray[0] ===0 :
            setExtraNewEndingPrincipalArray([principal]);
            break;
        case currentExtraPrincipal > monthlyPayment + extraPayment &&
        extraNewEndingPrincipalArray.length >= 1:
            processEachExtraPayment(1);
            break;
        case currentExtraPrincipal < monthlyPayment + extraPayment &&
        extraNewEndingPrincipalArray[extraNewEndingPrincipalArray.length - 1] > 0:
            processEachExtraPayment(0);
            break;
        case extraNewEndingPrincipalArray[
        extraNewEndingPrincipalArray.length - 1
            ] <= 0:
            break;
        default:
            break;
    }
};

export default ExtraPaymentCalculation;
