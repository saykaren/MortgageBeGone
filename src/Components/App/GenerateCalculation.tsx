import NumberConverter from './NumberConverter';

interface GenerateCalculationProps {
    newEndingPrincipalArray: number[];
    interestRate: number;
    monthlyPayment: number;
    principalPaidArray: number[];
    setPrincipalPaidArray: ([])=>void;
    interestPaidArray: number[];
    setInterestPaidArray: ([])=>void;
    setNewEndingPrincipalArray: ([])=>void;
    principal: number;
}

const GenerateCalculation = ({
                                 newEndingPrincipalArray,
                                 interestRate,
                                 monthlyPayment,
                                 principalPaidArray,
                                 setPrincipalPaidArray,
                                 interestPaidArray,
                                 setInterestPaidArray,
                                 setNewEndingPrincipalArray,
                                 principal,
                             }: GenerateCalculationProps) => {
    let currentPrincipal = 0;
    // let currentPrincipal;
    if (newEndingPrincipalArray.length > 0) {
        currentPrincipal = NumberConverter(
            newEndingPrincipalArray[newEndingPrincipalArray.length - 1],
        );
    }

    const processEachPayment = (num: number) => {
        let paymentInterestPaid = NumberConverter(
            currentPrincipal * ((interestRate * 0.01) / 12),
        );
        let principalPaid;
        ////0 is last payment
        if (num === 0) {
            principalPaid = NumberConverter(currentPrincipal);
        } else {
            principalPaid = NumberConverter(monthlyPayment - paymentInterestPaid);
        }
        let balance = NumberConverter(currentPrincipal - principalPaid);
        setPrincipalPaidArray([...principalPaidArray, principalPaid]);
        setInterestPaidArray([...interestPaidArray, paymentInterestPaid]);
        setNewEndingPrincipalArray([...newEndingPrincipalArray, balance]);
    };

    switch (true) {
        case newEndingPrincipalArray[0] === 0:
            setNewEndingPrincipalArray([principal]);
            break;
        case currentPrincipal > monthlyPayment &&
        newEndingPrincipalArray.length >= 1:
            processEachPayment(1);
            break;
        case currentPrincipal < monthlyPayment:
            processEachPayment(0);
            break;
        case newEndingPrincipalArray[newEndingPrincipalArray.length - 1] <= 0:
            break;
        default:
            break;
    }
};

export default GenerateCalculation;
