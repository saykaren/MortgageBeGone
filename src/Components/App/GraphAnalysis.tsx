import React from 'react';

interface GraphAnalysisProps {
  totalPaidData: number;
  graphPWidth: number;
  graphIWidth: number;
}

const GraphAnalysis = ({
  totalPaidData,
  graphPWidth,
  graphIWidth,
}: GraphAnalysisProps) => {
  return (
    <div className="dataSection">
      <div className="negative">Total Paid ${totalPaidData.toFixed(2)}</div>
      <div className="growGraph">
        <div
          id="principalBar"
          className="detailsColumn"
          style={{
            height: `${graphPWidth / 2}%`,
            backgroundColor: '#61dafb',
            border: '2px solid white',
            color: '#282c34',
          }}
        >
          {graphPWidth.toFixed(2)}%
        </div>
        <div
          id="interestPaid"
          className="detailsColumn"
          style={{
            height: `${graphIWidth / 2}%`,
            backgroundColor: 'red',
            border: '2px solid white',
            color: '#282c34',
          }}
        >
          {graphIWidth.toFixed(2)}%
        </div>
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-around' }}>
        <div>Principal</div>
        <div className="negative">Interest</div>
      </div>
    </div>
  );
};

export default GraphAnalysis;
