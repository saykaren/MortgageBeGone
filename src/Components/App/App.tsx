import React, {useState} from 'react';
import logo from '../Assests/logo.svg';
import '../Styling/App.scss';

function App() {
    const [counter, setCounter] = useState(0);

    return (
        <div className="App">
            <header className="App-header">
                <img src={logo} className="App-logo" alt="logo" />
                <p>
                    Editing <code>src/App.js</code> and save to reload.
                </p>
                <a
                    className="App-link"
                    href="https://reactjs.org"
                    target="_blank"
                    rel="noopener noreferrer"
                >
                    Learn React
                </a>
                <Counter counter={counter}/>
            </header>
        </div>
    );
}

interface CounterProps {
    counter: number;
}

export const Counter = ({counter}:CounterProps) =>{
return(
    <div>
        <p>{counter}</p>
    </div>
)
};

export default App;
