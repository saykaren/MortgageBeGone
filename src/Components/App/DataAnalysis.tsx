import React from 'react';
import TotalAmountAnalysis from './TotalAmountAnalysis';
import GraphAnalysis from './GraphAnalysis';
import TimeAnalysis from './TimeAnalysis';

interface DataAnaylsisProps {
    interestPaidArray: [];
    mortgage: number;
    extraInterestPaidArray: [];
    extraTotalPaidToBank: number;
    totalPaidToBank: number;
}

const DataAnalysis = ({
                          interestPaidArray,
                          mortgage,
                          extraInterestPaidArray,
                          extraTotalPaidToBank,
                          totalPaidToBank,
                      }: DataAnaylsisProps) => {
    const totalPaid = totalPaidToBank + mortgage;
    const extraTotalPaid = extraTotalPaidToBank + mortgage;
    let interestWidth = (totalPaidToBank / totalPaid) * 100;
    let principalWidth = 100 - interestWidth;

    let extraInterestWidth = (extraTotalPaidToBank / extraTotalPaid) * 100;
    let extraPrincipalWidth = 100 - extraInterestWidth;

    return (
        <section>
            {extraInterestPaidArray.length > 0 && (
                <div className="dataForm">
                    <h2 className="headerCalc">
                        Amortization with <span className="positive">Extra</span> Payment
                    </h2>
                    <div className="dataSection">
                        <TimeAnalysis
                            title="Time til loan paid off"
                            arr={extraInterestPaidArray}
                        />
                    </div>
                    <TotalAmountAnalysis
                        loan={mortgage}
                        principalWidth={extraPrincipalWidth}
                        intTotal={extraTotalPaidToBank}
                        interestWidth={extraInterestWidth}
                        overallTotal={extraTotalPaid}
                    />
                    <GraphAnalysis
                        totalPaidData={extraTotalPaid}
                        graphPWidth={extraPrincipalWidth}
                        graphIWidth={extraInterestWidth}
                    />
                </div>
            )}
            {interestPaidArray.length > 0 && (
                <div className="dataForm">
                    <h2 className="headerCalc">Amortization Typical Payment</h2>
                    <div className="dataSection">
                        <TimeAnalysis
                            title="Time til loan paid off"
                            arr={interestPaidArray}
                        />
                    </div>
                    <TotalAmountAnalysis
                        loan={mortgage}
                        principalWidth={principalWidth}
                        intTotal={totalPaidToBank}
                        interestWidth={interestWidth}
                        overallTotal={totalPaid}
                    />
                    <GraphAnalysis
                        totalPaidData={totalPaid}
                        graphPWidth={principalWidth}
                        graphIWidth={interestWidth}
                    />
                </div>
            )}
        </section>
    );
};

export default DataAnalysis;
