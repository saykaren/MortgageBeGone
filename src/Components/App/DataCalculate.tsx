import React, { useState, useEffect, ChangeEvent, FC } from 'react';
import NumberConverter from './NumberConverter';
import DataAnalysis from './DataAnalysis';
import GenerateCalculation from './GenerateCalculation';
import ExtraPaymentCalculation from './ExtraPaymentCalculation';
import AmortizationColumnSetValue from './AmortizationColumnSetValue';
import AmortizationColumn from './AmortizationColumn';
import InputLabelOne from "./InputLabelOne";
import Modal from './Modal';

const DataCalculate: React.FC = () => {
    // User Input
    const [principal, setPrincipal] = useState<number>(172000);
    const [interestRate, setInterestRate] = useState<number>(3.75);
    const [monthlyPayment, setMonthlyPayment] = useState<number>(1500);
    const [extraPayment, setExtraPayment] = useState<number>(1000);

    //Extra payment
    const [extraPrincipalPaidArray, setExtraPrincipalPaidArray] = useState<[] | null>([]);
    const [extraInterestPaidArray, setExtraInterestPaidArray] = useState<[] | null>([]);
    const [
        extraNewEndingPrincipalArray,
        setExtraNewEndingPrincipalArray,
    ] = useState<[] | null>([]);

    //Update Input
    const [principalPaidArray, setPrincipalPaidArray] = useState<[] | null>([]);
    const [interestPaidArray, setInterestPaidArray] = useState<[] | null>([]);
    const [newEndingPrincipalArray, setNewEndingPrincipalArray] = useState<[] | any[]>([]);

    // Modal
    const [modal, setModal] = useState<boolean>(false);

    //Calculations
    interface handleGenerateCalculationProps {
        newEndingPrincipalArray: [],
        interestRate: number,
        monthlyPayment: number,
        principalPaidArray: number[],
        setPrincipalPaidArray: ([])=>void,
        interestPaidArray: number[],
        setInterestPaidArray: number[],
        setNewEndingPrincipalArray: ([])=>void,
        principal: number,
    }

    const handleGenerateCalculation = () => {
        GenerateCalculation({
            newEndingPrincipalArray,
            interestRate,
            monthlyPayment,
            principalPaidArray,
            setPrincipalPaidArray,
            interestPaidArray,
            setInterestPaidArray,
            setNewEndingPrincipalArray,
            principal,
        });
    };

    const handleGenerateExtraCalculation = () => {
        ExtraPaymentCalculation({
            extraNewEndingPrincipalArray,
            interestRate,
            setModal,
            extraPayment,
            monthlyPayment,
            extraPrincipalPaidArray,
            setExtraPrincipalPaidArray,
            extraInterestPaidArray,
            setExtraInterestPaidArray,
            setExtraNewEndingPrincipalArray,
            principal,
        });
    };

    const handleReset = (e: string, setEvent: (arg1:number)=>void) => {
        let result = parseFloat(e);
        setEvent(result);
    };

    useEffect(() => {
        if (
            newEndingPrincipalArray.length >= 1 &&
            newEndingPrincipalArray[newEndingPrincipalArray.length - 1] > 0
        ) {
            setTimeout(handleGenerateCalculation, 100);
        }
    }, [newEndingPrincipalArray]);

    useEffect(() => {
        if (
            extraNewEndingPrincipalArray.length >= 1 &&
            extraNewEndingPrincipalArray[extraNewEndingPrincipalArray.length - 1] > 0
        ) {
            setTimeout(handleGenerateExtraCalculation, 100);
        }
    }, [extraNewEndingPrincipalArray]);

    const calculate = () => {
        handleGenerateCalculation();
        if (extraPayment > 0) {
            handleGenerateExtraCalculation();
        }
    };

    //Calculate total paid
    let totalPaidToBank = 0;
    let extraTotalPaidToBank = 0;

    if (interestPaidArray.length > 0) {
        // totalPaidToBank = NumberConverter(
        //     interestPaidArray.reduce((accu, cur) => accu + cur)
        // );
        totalPaidToBank = interestPaidArray.reduce(function(acc,num){
            return acc+num;}, 0);
    }
    if (extraInterestPaidArray.length > 0) {
        // extraTotalPaidToBank = NumberConverter(
        //     extraInterestPaidArray.reduce((accu, cur) => accu + cur),
        // );
        extraTotalPaidToBank = extraInterestPaidArray.reduce(function(acc,num){
            return acc+num;}, 0);
    }

    let savedYears = Math.floor(
        (interestPaidArray.length - extraInterestPaidArray.length) / 12,
    );
    let savedMonths =
        (interestPaidArray.length - extraInterestPaidArray.length) % 12;

    let savedTotalPaid = Math.floor(totalPaidToBank - extraTotalPaidToBank);

//     interface InputLabelOneProps {
//     handleReset: (arg1:string, arg2:(arg1:number)=>void)=>void;
//     title: string;
//     arrayCheck: [];
//     value: string;
//     setProperty: (arg1:number)=>void;
// }
//     const InputLabelOne = ({
//                                handleReset,
//                                title,
//                                arrayCheck,
//                                value,
//                                setProperty,
//                            }: InputLabelOneProps) => {
//         return (
//             <label className="inputSection">
//                 {title}:
//                 {(arrayCheck.length <= 1 && (
//                     <input
//                         type="number"
//                         name="Mortgage"
//                         value={value}
//                         onChange={(e) => handleReset(e.currentTarget.value, setProperty)}
//                     ></input>
//                 )) || <span>{value}</span>}
//             </label>
//         );
//     };

    return (
        <section className="App">
            {modal && (
                <Modal
                    savedYears={savedYears}
                    savedMonths={savedMonths}
                    setModal={setModal}
                    savedTotalPaid={savedTotalPaid}
                />
            )}
            <div id="inputSection">
                <form className="inputForm">
                    <InputLabelOne handleReset={handleReset} title={'Mortgage Checking'} arrayCheck={newEndingPrincipalArray} value={principal} setProperty={setPrincipal}/>
                    <InputLabelOne handleReset={handleReset} title={'Interest Rate'} arrayCheck={newEndingPrincipalArray} value={interestRate} setProperty={setInterestRate}/>
                    <InputLabelOne handleReset={handleReset} title={'Monthly Payment'} arrayCheck={newEndingPrincipalArray} value={monthlyPayment} setProperty={setMonthlyPayment}/>
                    <InputLabelOne handleReset={handleReset} title={'Extra Monthly Payment'} arrayCheck={newEndingPrincipalArray} value={extraPayment} setProperty={setExtraPayment}/>
                </form>
            </div>
            {(newEndingPrincipalArray[newEndingPrincipalArray.length - 1] < 0 ||
                newEndingPrincipalArray[newEndingPrincipalArray.length - 1] ===
                undefined) && <button onClick={() => calculate()}>Calculate</button>}
            <button onClick={() => window.location.reload()}>Reset Numbers</button>

            <DataAnalysis
                interestPaidArray={interestPaidArray}
                mortgage={principal}
                extraInterestPaidArray={extraInterestPaidArray}
                extraTotalPaidToBank={extraTotalPaidToBank}
                totalPaidToBank={totalPaidToBank}
            />

            {extraNewEndingPrincipalArray.length > 1 && (
                <div id="flexTable">
                    <AmortizationColumn title='Extra Principal' initialRow={principal} arr={extraNewEndingPrincipalArray}/>
                    <AmortizationColumnSetValue title='Monthly Payment' copyValue={monthlyPayment} arr={extraNewEndingPrincipalArray}/>
                    <AmortizationColumn title='Interest Paid (EXTRA Calculation)' initialRow='-' arr={extraInterestPaidArray}/>
                    <AmortizationColumn title='Principal Paid (EXTRA Calculation)' initialRow='-' arr={extraPrincipalPaidArray}/>
                    <AmortizationColumn title='Ending Principal (EXTRA Calculation)' initialRow='' arr={extraNewEndingPrincipalArray}/>
                </div>
            )}

            <div id="flexTable">
                <AmortizationColumn title='Principal' initialRow={principal} arr={newEndingPrincipalArray}/>
                <AmortizationColumnSetValue title={'Monthly Payment'} copyValue={monthlyPayment} arr={newEndingPrincipalArray}/>
                <AmortizationColumn title='Interest Paid' initialRow='-' arr={interestPaidArray} />
                <AmortizationColumn title='Principal Paid' initialRow='-' arr={principalPaidArray} />
                <AmortizationColumn title='Ending Principal' initialRow='' arr={newEndingPrincipalArray}/>
            </div>
        </section>
    );
};

export default DataCalculate;
