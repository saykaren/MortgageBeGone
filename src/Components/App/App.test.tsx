import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import renderer from 'react-test-renderer';

import {Counter} from './App';

describe('App check', ()=>{
    test('renders learn react link', () => {
        const { getByText } = render(<App />);
        const linkElement = getByText(/learn react/i);
        expect(linkElement).toBeInTheDocument();
    });

    // test('DataCalculation component present', () => {
    //
    //     const {getByTestId} = render(<App />);
    //     const elelment = getByTestId(document.getElementById('.App-link'));
    //     expect(elelment).toContain('.App-link');
    // });

    it('My test suite', ()=>{
        expect(true).toEqual(true);
    });

    test('App', ()=>{
        const component = renderer.create(<App/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot;
    });

    test('snapshot renders', ()=>{
        const component = renderer.create(<Counter counter={1} />);
        let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        })
    });

