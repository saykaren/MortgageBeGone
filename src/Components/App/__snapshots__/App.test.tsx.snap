// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App check snapshot renders 1`] = `
<div>
  <p>
    1
  </p>
</div>
`;
